src/lib/state-operators/adding.ts

```diff
@@ -0,0 +1,45 @@
+import { StateOperator } from '@ngxs/store';
+import { asArray } from '../internal';
+import { EntityStateModel, IdProvider } from '../models';
+
+/**
+ * Adds or replaces the given entities to the state.
+ * For each entity an ID will be calculated, based on the given provider.
+ * This operator ensures that the calculated ID is added to the entity, at the specified id-field.
+ * The `lastUpdated` timestamp will be updated.
+ * @param entities the new entities to add or replace
+ * @param idKey key of the id-field of an entity
+ * @param idProvider function to provide an ID for the given entity
+ */
+export function addOrReplace<T>(
+  entities: T | T[],
+  idKey: string,
+  idProvider: IdProvider<T>
+): StateOperator<EntityStateModel<T>> {
+  return (state: EntityStateModel<T>) => {
+    const nextEntities = { ...state.entities };
+    const nextIds = [...state.ids];
+    let nextState = state; // will be reassigned while looping over new entities
+
+    asArray(entities).forEach(entity => {
+      const id = idProvider(entity, nextState);
+      entity[idKey] = id; // ensure ID is in the entity
```

Comment:
This line will fail if the entity already exists in the state and we are in `developmentMode` because the object would have been frozen. You should check to see if the entity's id has changed and if it has then you should make a copy of it before modifying it (the `entities` are passed into this function so you therefore should not be modifying them because they do not belong to you... we want this function to be a pure function).

Reply:
Fixed the bug for `developmentMode`. The function should now be pure.

The integration app is already in `developmentMode`. Not sure why it didn't pop up.